mixnet/src/node.rs

```diff
@@ -1,27 +1,126 @@
+use rand::rngs::OsRng;
 use serde::{Deserialize, Serialize};
+use sphinx_packet::crypto::{PrivateKey, PRIVATE_KEY_SIZE};
 use tokio::sync::mpsc;
 
-use crate::{error::MixnetError, packet::Packet, packet::PacketBody};
+use crate::{
+    error::MixnetError,
+    fragment::{Fragment, MessageReconstructor},
+    packet::{Message, Packet, PacketBody},
+    poisson::Poisson,
+};
 
 /// Mix node implementation that returns Sphinx packets which needs to be forwarded to next mix nodes,
 /// or messages reconstructed from Sphinx packets delivered through all mix layers.
-pub struct MixNode {}
+pub struct MixNode {
+    output_rx: mpsc::UnboundedReceiver<Output>,
+}
+
+struct MixNodeRunner {
+    _config: MixNodeConfig,
+    encryption_private_key: PrivateKey,
+    poisson: Poisson,
+    packet_queue: mpsc::Receiver<PacketBody>,
+    message_reconstructor: MessageReconstructor,
+    output_tx: mpsc::UnboundedSender<Output>,
+}
 
 /// Mix node configuration
 #[derive(Serialize, Deserialize, Clone, Debug)]
-pub struct MixNodeConfig {}
+pub struct MixNodeConfig {
+    /// Private key for decrypting Sphinx packets
+    pub encryption_private_key: [u8; PRIVATE_KEY_SIZE],
+    /// Poisson delay rate per minutes
+    pub delay_rate_per_min: f64,
+}
+
+const PACKET_QUEUE_SIZE: usize = 256;
 
 /// Queue for sending packets to [`MixNode`]
 pub type PacketQueue = mpsc::Sender<PacketBody>;
 
 impl MixNode {
-    pub fn new(_: MixNodeConfig) -> Result<(Self, PacketQueue), MixnetError> {
-        todo!()
+    /// Creates a [`MixNode`] and a [`PacketQueue`].
+    ///
+    /// This returns [`MixnetError`] if the given `config` is invalid.
+    pub fn new(config: MixNodeConfig) -> Result<(Self, PacketQueue), MixnetError> {
+        let encryption_private_key = PrivateKey::from(config.encryption_private_key);
+        let poisson = Poisson::new(config.delay_rate_per_min)?;
+        let (packet_tx, packet_rx) = mpsc::channel(PACKET_QUEUE_SIZE);
+        let (output_tx, output_rx) = mpsc::unbounded_channel();
+
+        MixNodeRunner {
+            _config: config,
+            encryption_private_key,
+            poisson,
+            packet_queue: packet_rx,
+            message_reconstructor: MessageReconstructor::new(),
+            output_tx,
+        }
+        .run();
+
+        Ok((Self { output_rx }, packet_tx))
     }
 
     /// Returns a next `[Output]` to be emitted, if it exists and the Poisson delay is done (if necessary).
     pub async fn next(&mut self) -> Option<Output> {
-        todo!()
+        self.output_rx.recv().await
+    }
+}
+
+impl MixNodeRunner {
+    fn run(mut self) {
+        tokio::spawn(async move {
+            loop {
+                if let Some(packet) = self.packet_queue.recv().await {
+                    if let Err(e) = self.process_packet(packet) {
+                        tracing::error!("failed to process packet. skipping it: {e}");
+                    }
+                }
+            }
+        });
+    }
```

Comment:
I would make this method async directly. Reason is that `spawn`, imo, its better read explicitly.

```rust
tokio::spawn(mixnode_runner.run)
```

Reply:
Yeah, right. Thank you! 7e2ce5790e4bdac9f6d91858badfc24c48ea55a1